File: LeetCode/543.py
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        diameter = 0
        def DFS(current):
            nonlocal diameter
            if not current: return 0
            left = DFS(current.left)
            right = DFS(current.right)
            diameter = max(diameter, left + right + 1)
            return max(left, right) + 1
        
        DFS(root)
        return diameter - 1

File: LeetCode/543.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int ans{0};
    
    int maxDepth(TreeNode* root) {
        if(root == nullptr) return 0;
        auto left = maxDepth(root->left);
        auto right = maxDepth(root->right);
        ans = max(ans, left+right);
        return 1+max(left,right);
    }
    
public:
    int diameterOfBinaryTree(TreeNode* root) {
        maxDepth(root);
        return ans;
    }
};